test/subcommand_flags.cc

```diff
@@ -0,0 +1,314 @@
+      " a 'oneof subcommand' field");
+}
+
+TEST(FlagsTest, IncorrectOneofName) {
+  EXPECT_DEATH(
+      []() {
+        test::IncorrectOneofName flag;
+        auto builder = stout::flags::Parser::Builder(&flag);
+        builder.Build();
+      }(),
+      "'oneof' field must have 'subcommand' name. "
+      "Other names are illegal");
+}
+
+TEST(FlagsTest, SubcommandFlagExtension) {
+  EXPECT_DEATH(
+      []() {
+        test::SubcommandFlagExtension flag;
+        auto builder = stout::flags::Parser::Builder(&flag);
+        builder.Build();
+      }(),
+      "Every field of the 'oneof subcommand' must have"
+      " .stout.v1.subcommand. extension");
+}
+
+TEST(FlagsTest, SimpleSubcommandBuildSucceed) {
+  test::SimpleSubcommandSucceed flags;
+
+  auto parser = stout::flags::Parser::Builder(&flags).Build();
+
+  std::array arguments = {
+      "program",
+      "--b",
+      "build",
+      "--other_flag=13",
+  };
+
+  int argc = arguments.size();
+  const char** argv = arguments.data();
+
+  parser.Parse(&argc, &argv);
+
+  EXPECT_EQ(1, argc);
+  EXPECT_EQ("program", argv[0]);
+  EXPECT_TRUE(flags.b());
+  EXPECT_EQ(13, flags.build().other_flag());
+}
+
+TEST(FlagsTest, SimpleSubcommandInfoSucceed) {
+  test::SimpleSubcommandSucceed flags;
+
+  auto parser = stout::flags::Parser::Builder(&flags).Build();
+
+  std::array arguments = {
+      "program",
+      "--b",
+      "info_flag",
+      "--info=hello world",
+  };
+
+  int argc = arguments.size();
+  const char** argv = arguments.data();
+
+  parser.Parse(&argc, &argv);
+
+  EXPECT_EQ(1, argc);
+  EXPECT_EQ("program", argv[0]);
+  EXPECT_TRUE(flags.b());
+  EXPECT_EQ("hello world", flags.info_flag().info());
+}
+
+TEST(FlagsTest, DuplicateSubcommands) {
+  test::SimpleSubcommandSucceed flags;
+
+  auto parser = stout::flags::Parser::Builder(&flags).Build();
+
+  std::array arguments = {
+      "program",
+      "--b",
+      "info_flag",
+      "--info=hello world",
+      "info_flag",
+      "--info=oops",
+  };
+
+  int argc = arguments.size();
+  const char** argv = arguments.data();
+
+  EXPECT_DEATH(
+      parser.Parse(&argc, &argv),
+      "Encountered duplicate subcommand 'info_flag'");
+}
+
+TEST(FlagsTest, DuplicateSubcommandFlags) {
+  test::SimpleSubcommandSucceed flags;
+
+  auto parser = stout::flags::Parser::Builder(&flags).Build();
+
+  std::array arguments = {
+      "program",
+      "--b",
+      "build",
+      "--other_flag=1",
+      "--other_flag=2",
+  };
+
+  int argc = arguments.size();
+  const char** argv = arguments.data();
+
+  EXPECT_DEATH(
+      parser.Parse(&argc, &argv),
+      "Encountered duplicate flag 'other_flag'");
+}
+
+TEST(FlagsTest, DuplicateSubcommandFlagNameForEnclosingLevel) {
+  test::DuplicateEnclosingFlagName flags;
+
+  auto parser = stout::flags::Parser::Builder(&flags).Build();
+
+  std::array arguments = {
+      "program",
+      "--other_flag=1",
+      "build",
+      "--other_flag=2",
```

Comment:
We can support this case if we actually store nested parsers, see other comment. Let's do that!